In [73]:
#0.1 importing Labraries
import pandas as pd
import numpy as pn 
import os

In [74]:
df_prods = pd.read_csv(os.path.join('C:\\Users\\bukola\\Desktop\\Lanre Comp\\Documents\\CF\\Instacart Basket Analysis', '02 Data', '4.3_orders_products', 'products.csv'), index_col = False)

In [75]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [76]:
df_ords = pd.read_csv(os.path.join('C:\\Users\\bukola\\Desktop\\Lanre Comp\\Documents\\CF\\Instacart Basket Analysis', '02 Data', 'orders_wrangled.csv'), index_col = False)

In [87]:
df_ords.head(20)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   1187899        1            11                   4                  8   
11   2168274        2             1                   2                 11   
12   1501582        2             2                   5                 10   
13   1901567        2             3                   1                 10   
14    738281        2             4                   2                 10   
15   1673511        2             5                   3                 11   
16   1199898        2             6                   2                  9   
17   3194192        2             7                   2                 12   
18    788338        2             8                   1                 15   
19   1718559        2             9                   2                  9   

    days_since_prior_order  
0                      NaN  
1                     15.0  
2                     21.0  
3                     29.0  
4                     28.0  
5                     19.0  
6                     20.0  
7                     14.0  
8                      0.0  
9                     30.0  
10                    14.0  
11                     NaN  
12                    10.0  
13                     3.0  
14                     8.0  
15                     8.0  
16                    13.0  
17                    14.0  
18                    27.0  
19                     8.0

# Describe function on  Df_ords data Frame

In [78]:
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# The result genereated from the describe fuction shows that the data looks good taken for example the out of the manimum and maximum of days in a week in the "orders_day_of_week is 0 and 6 respectively amd also for order_hour_of_day is between 0hrs to 23hrs which doest not exced the 24hrs in a day.  


# checkng for mixed data

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

In [79]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

# There is no mix data in the data frame 

*Checking for Missing values in  df_ords*

==>df_ords.isnull().sum()

In [80]:
df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

 **There are 206209 data missing in the "days_since_prior_order''** the missig values are numeric. since is numeric am going to update the column with either mean or the meadian of the data statistics

In [83]:
df_nan1 = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [84]:
df_nan1

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

#Replacing the NAN in the days_since_prior_day missing data with the mean Value. The mean valus is 11
df_ords['days_since_prior_order'].fillna(mean, inplace=True)

In [88]:
df_ords['days_since_prior_order'].fillna(11, inplace=True)

In [89]:
df_ords.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  
0                    11.0  
1                    15.0  
2                    21.0  
3                    29.0  
4                    28.0

# 
order_id 2539329 has 'Nan' value before now with mean value of 11


In [92]:
df_ords.shape


(3421083, 6)

# checking for duplication 

In [94]:
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [95]:
df_ords.shape

(3421083, 6)

In [97]:
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [100]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

** There is no duplicate in the data sets**

# checking missing data in df_prods dataframe 

==> df_prods.isnull().sum()

In [61]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [43]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [70]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [71]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [48]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [50]:
df_prods_clean.shape

(49677, 5)

In [101]:
df_prods.dropna(inplace = True)

In [104]:
df_dups2 = df_prods_clean[df_prods_clean.duplicated()]

In [105]:
df_dups2

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [106]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [107]:
df_prods_clean_no_dups.shape

(49672, 5)

Data Frame (Clean) for df_ords and df_prod

In [116]:
df_ords.to_csv(os.path.join('C:\\Users\\bukola\\Desktop\\Lanre Comp\\Documents\\CF\\Instacart Basket Analysis\\orders_wrangled.csv'), index= False)

In [115]:
df_prods.to_csv(os.path.join('C:\\Users\\bukola\\Desktop\\Lanre Comp\\Documents\\CF\\Instacart Basket Analysis\\products_clean.csv'), index = False)